### DISCUSSION TITLE
quill:: logger() returns null in example_trival ??

### DISCUSSION THREAD

**Discussion by mbhardwa** [odygrd/quill#235] [Q&A]
hi - is Quill supported on Centos 7 & with gcc 7. 3.1. I run the compiler with std=c++17
It builds fine but the example_trival crashes since the default logger returned is null.
and if try to run examples for file loggers they hang on the a pthread_mutex in the quill::file_handler call.
Seems like some kinda build issue so wondering if people have seen such issues

**Comment by odygrd** [comment]
hey, that sounds strange

Can you please try to remove this whole line https://github.com/odygrd/quill/blob/master/examples/example_trivial.cpp#L17 and let me know if it still doesn't work ?

**Comment by odygrd** [comment]
I have tried to compile it and run it on ubuntu with gcc 7.5.0 and everything compiles (except 1 test) and the example runs okay. The default logger gets initialised when `quill:start()` is called and it should be initialised when that function returns ..

**Comment by odygrd** [comment]
Hello, I have created a centos 7 instance on AWS ( https://aws.amazon.com/marketplace/pp/prodview-qkzypm3vjr45g )

I installed to this instance devtoolset-7 (https://www.softwarecollections.org/en/scls/rhscl/devtoolset-7/) and CMake.

I tried to run the examples and I got a `std::bad_alloc` exception, it seems that gcc 7.3.1 didn't like that the `static` keyword was missing from some arrays. 

I have pushed the fixes to master and I have also fixed the compiler warnings. The examples seem to run fine on my centos 7 image now.

Those errors are looking different from the ones you reported but give it a try with the latest master and let me know if you are still having issues please.

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/7397786/214680550-44e82b78-3996-4bbd-b7e4-c25d0ba57a80.png">

**Comment by mbhardwa** [comment]
Thanks for looking into this.
So after your changes the examples run
But ( this is super strange)
When I copy the code and run separately the
Quill::rotating-file-handler(name, 1024,5)
call just hangs in pthread mutex lock
But the examples work
I only have quill:: start before call to file handler.
If instead I just get default logger that works
Btw if I turn on use bounded que flag, the simple logger crashes in my code
Again gcc 7.3.1 with std =c++17

Sent from my iPhone

On Jan 25, 2023, at 3:23 PM, Odysseas Georgoudis ***@***.***> wrote:

﻿

Hello, I have created a centos 7 instance on AWS ( https://aws.amazon.com/marketplace/pp/prodview-qkzypm3vjr45g )

I installed to this image devtoolset-7 (https://www.softwarecollections.org/en/scls/rhscl/devtoolset-7/) and CMake.

I tried to run the examples and was getting a std::bad_aloc exception, it seems that gcc 7.3.1 didn't like that the static keyword was missing from some arrays.

I have pushed the fixes to master and I have also fixed the compiler warnings. The examples seem to run fine on my centos 7 image now.

Those errors are looking different from the ones you reported but give it a try with the latest master and let me know if you are still having issues please.

[image]<https://user-images.githubusercontent.com/7397786/214680550-44e82b78-3996-4bbd-b7e4-c25d0ba57a80.png>

**Comment by mbhardwa** [comment]
I figured out my issue.
Was not including -DQull-no-exceptions flag in my own build where as quill was built with flag on.

It runs now, thanks for your help

Sent from my iPhone

On Jan 25, 2023, at 5:51 PM, Manish BHARDWAJ ***@***.***> wrote:

﻿ Thanks for looking into this.
So after your changes the examples run
But ( this is super strange)
When I copy the code and run separately the
Quill::rotating-file-handler(name, 1024,5)
call just hangs in pthread mutex lock
But the examples work
I only have quill:: start before call to file handler.
If instead I just get default logger that works
Btw if I turn on use bounded que flag, the simple logger crashes in my code
Again gcc 7.3.1 with std =c++17

Sent from my iPhone

On Jan 25, 2023, at 3:23 PM, Odysseas Georgoudis ***@***.***> wrote:

﻿

Hello, I have created a centos 7 instance on AWS ( https://aws.amazon.com/marketplace/pp/prodview-qkzypm3vjr45g )

I installed to this image devtoolset-7 (https://www.softwarecollections.org/en/scls/rhscl/devtoolset-7/) and CMake.

I tried to run the examples and was getting a std::bad_aloc exception, it seems that gcc 7.3.1 didn't like that the static keyword was missing from some arrays.

I have pushed the fixes to master and I have also fixed the compiler warnings. The examples seem to run fine on my centos 7 image now.

Those errors are looking different from the ones you reported but give it a try with the latest master and let me know if you are still having issues please.

[image]<https://user-images.githubusercontent.com/7397786/214680550-44e82b78-3996-4bbd-b7e4-c25d0ba57a80.png>

**Comment by odygrd** [comment]
By default the library builds with exceptions on, if you are building without exceptions you need to pass `add_definitions(-DQUILL_NO_EXCEPTIONS)`. In that case exceptions are converted to assertions.

The library also needs to link against `Threads` https://github.com/odygrd/quill/blob/master/quill/CMakeLists.txt#L151 but I guess you have already figured it out.

Since you are using the BoundedQueue if you want to change the size of the queue you can do it by calling `quill::configure` before `quill::start` 

See https://github.com/odygrd/quill/blob/master/quill/include/quill/Config.h#L182 for all the runtime config options.

for example

```cpp
  quill::Config cfg;
  cfg.default_queue_capacity = 131'072;
  quill::configure(cfg);
  quill::start();
```

Also note that if the bounded queue gets full, log messages will be dropped, in that case the library will log the number of dropped log messages to `stderr`

I have ran the whole test suite on `centos 7` with `gcc 7.3.1` and everything seems okay. Only one filesystem test failed which I have compiled out for now, it shouldn't be an issue https://github.com/odygrd/quill/commit/bcb1440b211cf9ccacca085e68a42291d2121c69

Current master will be `v2.7.0` I am planning to release that version in 2-4 weeks from today

Happy you got that working, if you find any other issues please let me know!

**Comment by mbhardwa** [comment]
A suggestion would be to add a note in documentation that if someone is turning on options( like no exceptions) they should compile their code ( which includes quill.h) with the appropriate -D flag
Otherwise looks pretty good till now. Will let you know if something else comes up. Thanks again

Sent from my iPhone

On Jan 25, 2023, at 7:46 PM, Odysseas Georgoudis ***@***.***> wrote:

﻿

By default the library builds with exceptions on, if you are building without exceptions you need to pass add_definitions(-DQUILL_NO_EXCEPTIONS). In that case exceptions are converted to assertions.

The library also needs to link against Threads https://github.com/odygrd/quill/blob/master/quill/CMakeLists.txt#L151 but I guess you have already figured it out.

Since you are using the BoundedQueue if you want to change the size of the queue you can do it by calling quill::configure before quill::start

See https://github.com/odygrd/quill/blob/master/quill/include/quill/Config.h#L182 for all the runtime config options.

for example

  quill::Config cfg;
  cfg.default_queue_capacity = 131'072;
  quill::configure(cfg);
  quill::start();

Also note that if the bounded queue gets full, log messages will be dropped, in that case the library will log the number of dropped log messages to stderr

I have ran the whole test suite on centos 7 with gcc 7.3.1 and everything seems okay. Only one filesystem test failed which I have compiled out for now, it shouldn't be an issue bcb1440<https://github.com/odygrd/quill/commit/bcb1440b211cf9ccacca085e68a42291d2121c69>

Happy you got that working, if you find any other issues please let me know!

**Comment by odygrd** [comment]
Thanks, I see what you mean, that is actually a bug in the CMakeLists. I have pushed a fix https://github.com/odygrd/quill/commit/f7327e78db6a4b4cae60fc9bef5fdc5e9eb6b867
